pennylane/spin/lattice.py

```diff
@@ -0,0 +1,196 @@
+
+        if len(self.boundary_condition) != self.n_dim:
+            raise ValueError(
+                "Argument 'boundary_condition' must be a bool or a list of bools of same dimensions as the unit_cell"
+            )
+
+    def _identify_neighbours(self, cutoff):
+        r"""Identifies the connections between lattice points and returns the unique connections
+        based on the neighbour_order"""
+
+        tree = cKDTree(self.lattice_points)
+        indices = tree.query_ball_tree(tree, cutoff)
+        unique_pairs = set()
+        edges = {}
+        for i, neighbours in enumerate(indices):
+            for neighbour in neighbours:
+                if neighbour != i:
+                    pair = (min(i, neighbour), max(i, neighbour))
+                    if pair not in unique_pairs:
+                        unique_pairs.add(pair)
+                        dist = np.linalg.norm(
+                            self.lattice_points[i] - self.lattice_points[neighbour]
+                        )
+                        # Scale the distance
+                        bin_density = 21621600  # multiple of expected denominators
+                        scaled_dist = np.rint(dist * bin_density)
+
+                        if scaled_dist not in edges:
+                            edges[scaled_dist] = []
+                        edges[scaled_dist].append((i, neighbour))
+        return edges
+
+    def _generate_true_edges(self, edges, map, neighbour_order):
+        r"""Modifies the edges to remove hidden nodes and create connections based on boundary_conditions"""
+
+        self.edges = []
+        for i, (_, edge) in enumerate(sorted(edges.items())):
+            if i >= neighbour_order:
+                break
+            for e1, e2 in edge:
+                true_edge = (min(map[e1], map[e2]), max(map[e1], map[e2]), i)
+                if true_edge not in self.edges:
+                    self.edges.append(true_edge)
+
+    def _generate_grid(self, neighbour_order):
+        """Generates the coordinates of all lattice sites and their indices.
+
+        Args:
+           neighbour_order: The number of nearest neighbour interactions.
+        Returns:
+           lattice_points: The coordinates of all lattice sites.
+           lattice_map: A list to represent the node number for each lattice_point
+        """
+
+        n_sl = len(self.basis)
+        if self.boundary_condition:
+            wrap_grid = np.where(self.boundary_condition, neighbour_order, 0)
+        else:
+            wrap_grid = np.zeros(self.L.size, dtype=int)
+
+        ranges_dim = [range(-wrap_grid[i], Lx + wrap_grid[i]) for i, Lx in enumerate(self.L)]
+        ranges_dim.append(range(n_sl))
+        lattice_points = []
+        lattice_map = []
+
+        for Lx in itertools.product(*ranges_dim):
+            point = np.dot(Lx[:-1], self.unit_cell) + self.basis[Lx[-1]]
+            node_index = 0
+            for i in range(self.n_dim):
+                node_index += (
```

Comment:
Maybe this could be vectorized (and made efficient) via `np.mod` and `np.cumprod`?